hebsafeharbor/identifier/signals/hebrew_city_recognizer.py

```diff
@@ -57,27 +59,138 @@ def analyze(
+
+                    # Update the explainability object with context information
+                    # helped improving the score
+                    result.analysis_explanation.set_supportive_context_word(
+                        supportive_context
+                    )
+                    result.analysis_explanation.set_improved_score(result.score)
+
+        return raw_recognizer_results
+
+    @staticmethod
+    def _extract_geo_entities(nlp_artifacts: NlpArtifacts, entity_names: List[str]) -> Set[int]:
+        """
+        Identify positions of other entities in a text that can provide context for entity
+        :param nlp_artifacts: artifacts of the nlp engine
+        :param entity_names: list of entity names
+        :return set of positions in a text string corresponding provided entity names
+        """
+        geo_entities_positions = set()
+        for entity in nlp_artifacts.entities:
+            if entity.label_ in entity_names:
+                for char in range(entity.start_char, entity.end_char):
+                    geo_entities_positions.add(char)
+        return geo_entities_positions
+
+    def enhance_using_other_entities(self, nlp_artifacts: NlpArtifacts, results: List[RecognizerResult]):
+        """
+        Iterate over the spaCy tokens, and find all the position indices for endorsing entities
+        :param nlp_artifacts: artifacts of the nlp engine
+        :param results: list of preliminary recognizer results
+        :return list of results, enhanced where there is an overlap with geo entity
+        """
+        geo_entities_positions = self._extract_geo_entities(nlp_artifacts, self.endorsing_entities)
+        if not geo_entities_positions:
+            return results
+
+        for result in results:
+            overlap_ratio = len(
+                set(range(result.start, result.end)).intersection(geo_entities_positions)) / (result.end - result.start)
+            adjusted_score = result.score + (overlap_ratio * self.LOCATION_OVERLAP_FACTOR)
+            if adjusted_score != result.score:
+                result.score = adjusted_score
+                result.analysis_explanation.set_improved_score(adjusted_score)
+                result.analysis_explanation.append_textual_explanation_line("NLP-LOC-enhanced;")
+
         return results
+
+    @staticmethod
+    def __extract_sentence_containing_position(nlp_artifacts: NlpArtifacts, start: int) -> str:
+        """
+        Extracting the sentence which contains the start position of recognized entity using sentences from nlp_artifacts
+        :param nlp_artifacts: An abstraction layer which holds different
+                              items which are the result of a NLP pipeline
+                              execution on a given text
+        :param start: The start index of the word in the original text
+        :return: sentence string
+        """
+        for sent in nlp_artifacts.tokens.sents:
```

Comment:
I believe this uses spaCy's sentencizer under the hood (maybe i'm wrong). Worth validating that it works well in Hebrew and in our specific use case.
https://spacy.io/usage/linguistic-features#sbd

Reply:
Anyway, this is super cool :) I like the use of the NLP Artifacts to go deeper

Reply:
In the examples I used it seemed to split quite well.